pyro/distributions/sine_skewed.py

```diff
@@ -0,0 +1,136 @@
+import warnings
+from math import pi
+
+import torch
+from torch import broadcast_shapes
+from torch.distributions import Uniform
+
+from pyro.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+
+
+class SineSkewed(TorchDistribution):
+    """Sine Skewing [1] is a procedure for producing a distribution that breaks pointwise symmetry on a torus
+    distribution. The new distribution is called the Sine Skewed X distribution, where X is the name of the (symmetric)
+    base distribution.
+
+    Torus distributions are distributions with support on products of circles
+    (i.e., ⨂^d S^1 where S^1=[-pi,pi) ). So, a 0-torus is a point, the 1-torus is a circle,
+    and the 2-torus is commonly associated with the donut shape.
+
+    The Sine Skewed X distribution is parameterized by a weight parameter for each dimension of the event of X.
+    For example with a von Mises distribution over a circle (1-torus), the Sine Skewed von Mises Distribution has one
+    skew parameter. The skewness parameters can be inferred using :class:`~pyro.infer.HMC` or :class:`~pyro.infer.NUTS`.
+    For example, the following will produce a uniform prior over skewness for the 2-torus,::
+
+        def model(obs):
+            # Sine priors
+            phi_loc = pyro.sample('phi_loc', VonMises(pi, 2.))
+            psi_loc = pyro.sample('psi_loc', VonMises(-pi / 2, 2.))
+            phi_conc = pyro.sample('phi_conc', Beta(halpha_phi, beta_prec_phi - halpha_phi))
+            psi_conc = pyro.sample('psi_conc', Beta(halpha_psi, beta_prec_psi - halpha_psi))
+            corr_scale = pyro.sample('corr_scale', Beta(2., 5.))
+
+            # SS prior
+            skew_phi = pyro.sample('skew_phi', Uniform(-1., 1.))
+            psi_bound = 1 - skew_phi.abs()
+            skew_psi = pyro.sample('skew_psi', Uniform(-1., 1.))
+            skewness = torch.stack((skew_phi, psi_bound * skew_psi), dim=-1)
+            assert skewness.shape == (num_mix_comp, 2)
```

Comment:
It would be nice to have a constraint+transform for this (in a follow-up PR). I believe we can use [signed stick-breaking transform](https://github.com/pytorch/pytorch/blob/e66015dadf135f6e1cda69d80883505807b1bbed/torch/distributions/transforms.py#L760) here. This way users can define distributions over `skewness` (or just simply use `pyro.param` with correct constraint). Without that, it is a bit cumbersome for users to define correct `skewness` over general d-torus.

Reply:
With that, we can have correct `constraints` in the distribution definition. :D

Reply:
That would be neat; I'll add it to the backlog.